pints/interfaces/_stan.py

```diff
@@ -0,0 +1,179 @@
+            raise RuntimeError(
+                "No data supplied to create Stan model fit object. " +
+                "Run `update_data` first.")
+        vals = self._prepare_values(x)
+        try:
+            val = self._log_prob(vals, adjust_transform=True)
+            dp = self._grad_log_prob(vals, adjust_transform=True)
+            return val, dp.reshape(-1)
+        except (RuntimeError, ValueError):
+            return -np.inf, np.ones(self._n_parameters).reshape(-1)
+
+    def _initialise(self, stanfit):
+        """ Initialises internal variables. """
+        if stanfit is None:
+            self._fit = None
+            self._log_prob = None
+            self._grad_log_prob = None
+            self._n_parameters = None
+            self._names = None
+            self._index = None
+            self._long_names = None
+            self._counter = None
+            self._dict = None
+        else:
+            self._fit = stanfit
+            self._log_prob = stanfit.log_prob
+            self._grad_log_prob = stanfit.grad_log_prob
+            names = stanfit.unconstrained_param_names()
+            self._n_parameters = len(names)
+            self._names, self._index = self._initialise_dict_index(names)
+            self._long_names = names
+            self._counter = Counter(self._index)
+            self._dict = {self._names[i]: [] for i in range(len(self._names))}
+
+    def _initialise_dict_index(self, names):
+        """ Initialises dictionary and index of names. """
+        names_short = []
+        for name in names:
+            num = name.find(".")
+            if num < 0:
+                names_short.append(name)
+            else:
+                names_short.append(name[:num])
+        names_long = list(names_short)
+        names_short = list(dict.fromkeys(names_short))
+        index = [names_short.index(name) for name in names_long]
+        return names_short, index
+
+    def names(self):
+        """ Returns names of Stan parameters. """
+        return self._long_names
+
+    def n_parameters(self):
+        """ See `pints.LogPDF.n_parameters`. """
+        return self._n_parameters
+
+    def update_data(self, stan_data):
+        """
+        Updates data passed to the underlying Stan model.
+
+        Parameters
+        ----------
+        stan_data
+            Data in Python dictionary format as required by PyStan.
+        """
+        stanfit = self._compiled_stan.sampling(
+            data=stan_data, iter=1, chains=1,
+            verbose=False, refresh=0,
+            control={'adapt_engaged': False})
+        self._initialise(stanfit)
+
+    def _prepare_values(self, x):
+        """ Flattens lists from PyStan's dictionary. """
+        dict = self._dict_update(x)
+        vals = dict.values()
+        b = []
+        for ele in vals:
+            if not isinstance(ele, list):
+                ele = [ele]
+            b.append(ele)
+        vals = [item for sublist in b for item in sublist]
```

Comment:
Is there a slightly more readable way to write this? I'm struggling a bit. But no worries if not :D